keps/sig-cli/3517-kubectl-client-side-apply-migration/README.md

```diff
@@ -0,0 +1,1113 @@
+migrating](https://kubernetes.io/docs/reference/using-api/server-side-apply/#upgrading-from-client-side-apply-to-server-side-apply) from usage of client-side-apply to server-side-apply:
+
+> Client-side apply users who manage a resource with kubectl apply can start using server-side apply with the following flag.
+>
+>     kubectl apply --server-side [--dry-run=server]
+
+
+Unfortunately it is not so simple. By following these instructions you can be
+left in a situation where after migrating to server-side-apply, certain fields
+become un-deletable from the object without a manual update repairing the managed fields. 
+Here is a simple reproducing case:
+
+Create a configmap in the cluster with client-side apply:
+```shell
+cat <<EOF | kubectl apply -f -
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+  legacy: unused
+EOF
+```
+```console
+configmap/test created
+```
+
+Apply the same manifest with `--server-side` flag, as per server-side-apply migration instructions:
+```shell
+cat <<EOF | kubectl apply --server-side -f -
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+  legacy: unused
+EOF
+```
+```console
+configmap/test serverside-applied
+```
+
+Remove one of the values from the configmap, we choose `legacy` key, and apply
+again using server-side apply. You would expect that the field is now removed
+from the object, but it is not:
+```shell
+cat <<EOF | kubectl apply --server-side -f -
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+EOF
+```
+```console
+configmap/test serverside-applied
+```
+
+Check the configmap, expecting the legacy value to be removed, and be surprised that it remains:
+```shell
+kubectl get configmap -o yaml test
+```
+```console
+apiVersion: v1
+kind: ConfigMap
+data:
+  key: value
+  legacy: unused
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
+      {"apiVersion":"v1","data":{"key":"value"},"kind":"ConfigMap","metadata":{"name":"test","namespace":"default"}}
+  name: test
+  namespace: default
+```
+
+This KEP is intended to fix this problem by providing users with a path to completely
+migrate their managed fields to be used with server-side-apply.
+
+### Goals
+
+This bug prevents users of server-side-apply, which is in GA, from deleting any fields of their objects unless they revert back to using client-side-apply. It especially poses huge
+challenges to cluster operators seeking to upgrade a large collection of resources to server-side-apply. Without an automatic method of preparing objects for use with SSA,
+adoption for server-side-apply is blocked for many users.
+
+1. The primary goal of this KEP is to provide a pathway for users blocked by this issue to adopt SSA.
+2. Provide an idempotent migration to convert large collections of resources to server-side apply, especially for automatic systems
```

Comment:
Also relevant is that SSA was recently made on-by-default for servers https://github.com/kubernetes/kubernetes/pull/112748

(For some reason it was still opt-in, despite being GA)

I think definitely `--server-side` on-by-default should be revisited. Personally I dont have context to say what use cases would be blocking such a change.

@apelisse maybe you have thoughts?

Reply:
Yes, the goal is to eventually remove CSA. We may have to wait for any bug fixes or support on the server to be universally available before we make it the default for kubectl? Aside from that I'm in favor of doing this as rapidly as possible :)

Reply:
A few important SSA bug fixes landed 1.24-1.25. And kubectl version skew is 1 minor version point. Given that SSA is to be enabled by default in 1.27 then the kubectl version for 1.28 will be the first that could enable `--server-side` by default.

Reply:
Having reflected on where this proposal has landed and the agreement we seem to have that the ultimate goal is to remove CSA, I have a suggestion. Unlike the changes we were talking about when this KEP started, the change that it is still proposing now feels a lot like a bug fix in the current migration implementation. The KEP process has provided a lot of value in landing us on this change and discussing the particulars. But unless there's a risk I'm not properly appreciating (certainly possible--I'm a newer lead and have not done all that much direct work on kubectl!), I'm not sure it actually makes sense to put a bug fix like this through the tracking and graduation process, significantly delaying availability to affected users. Yet, we have uncovered something bigger that does warrant careful multi-stage planning and communication with the community: that goal of removing CSA entirely. So far we have discussed the need for a transition process for the SSA flag, a blocking interaction with pruning, and several other kubectl commands that currently interact with the CSA annotation. So, my suggestion is that we take the bug fix currently described in this KEP and just do it in the next release, and pivot this KEP to plan that larger process, in which blocking bugs like this could be enumerated as part of the larger plan. Wdyt?

Reply:
@KnVerey I'm 100% on board with that plan.